lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/Evaluators.java

```diff
@@ -47,6 +44,10 @@
    */
   private final Map<String, EvaluatorManager> evaluators = new HashMap<>();
 
+  /**
+   * A map between evaluatorId and evaluator status for failed and returned evaluators.
+   */
+  private final Set<String> closedEvaluators = new HashSet<>();
```

Comment:
This should be a monotonic set. Turns out we have one in [Tang](https://github.com/apache/reef/blob/master/lang/java/reef-tang/tang/src/main/java/org/apache/reef/tang/util/MonotonicHashSet.java)